src/cmd/ksh93/sh/deparse.c

```diff
@@ -469,22 +469,17 @@ static_fn void p_comlist(const struct dolnod *dol, int endchar) {
     int flag = ' ', special;
 
     argv = dol->dolval + ARG_SPARE;
-    cp = *argv;
-    special = (cp[0] == '[' && cp[1] == 0);
+    cp = *argv++;
+    special = strcmp(cp, "[") == 0;
     do {
-        if (cp) {
-            argv++;
-        } else {
-            cp = "";
```

Comment:
What would happen if `cp` points to `NULL` character ?

Reply:
If you're asking what happens if `cp` is NULL that can't happen. At this juncture we've already dereferenced cp before entering the loop or we've exited the loop because `cp` is NULL (see the `while(...)` condition). Now it may be the case that statement `special = strcmp(cp, "[") == 0;` also needs to also check that `cp` is non-NULL. 

If you really do mean what happens if `cp` points to an empty string (first bye is `'\0'`) then the `cp = "";` doesn't change that situation.